A78-grpc-metrics-wrr-pf-xds.md

```diff
@@ -0,0 +1,154 @@
+
+### XdsClient
+
+The XdsClient component was originally described in [A27].  Note that in
+[A71], we are moving from a single global XdsClient instance to a
+separate global XdsClient instance for each channel target.  The
+proposed metric schema here reflects that change.
+
+XdsClient metrics will have the following labels:
+
+| Name        | Disposition | Description |
+| ----------- | ----------- | ----------- |
+| grpc.target | required | For clients, indicates the target of the gRPC channel in which the XdsClient is used (i.e., the same as the attribute defined in [A66]). For servers, will be the string "#server". |
+| grpc.xds.server | required | The name of the xDS server with which the XdsClient is communicating. |
+| grpc.xds.authority | required | The xDS authority.  The value will be "#old" for old-style non-xdstp resource names. |
+| grpc.xds.cache_state | required | Indicates the cache state of an xDS resource.  The value will be one of: <ul><li>"requested": The resource has been requested from the xDS server but has not yet been received.<li>"does_not_exist": The server has indicated that the resource does not exist.<li>"acked": The resource has been received and is valid.<li>"nacked": The resource was received but was not valid.<li>"nacked_but_cached": There is a version of the resource cached, but the most recent update of the resource was invalid.</ul> |
+| grpc.xds.resource_type | required | Indicates an xDS resource type, such as "envoy.config.listener.v3.Listener". |
+
+The following metrics will be exported:
+
+| Name          | Type  | Unit  | Labels  | Description |
+| ------------- | ----- | ----- | ------- | ----------- |
+| grpc.xds_client.connected | Gauge | {bool} | grpc.target, grpc.xds.server | Whether or not the xDS client currently has a working ADS stream to the xDS server.  For a given server, this will be set to 1 when the stream is initially created.  It will be set to 0 when we have a connectivity failure or when the ADS stream fails without seeing a response message, as per [A57].  Once set to 0, it will be reset to 1 when we receive the first response on an ADS stream. |
+| grpc.xds_client.resource_updates_valid | Counter | {resource} | grpc.target, grpc.xds.server, grpc.xds.resource_type | A counter of resources received that were considered valid.  The counter will be incremented even for resources that have not changed. |
+| grpc.xds_client.resource_updates_invalid | Counter | {resource} | grpc.target, grpc.xds.server, grpc.xds.resource_type | A counter of resources received that were considered invalid. |
+| grpc.xds_client.resources | Gauge | {resource} | grpc.target, grpc.xds.authority, grpc.xds.cache_state, grpc.xds.resource_type | Number of xDS resources. |
+
+### Temporary environment variable protection
+
+This proposal does not include any features enabled via external I/O, so
```

Comment:
Well, but we also don't want them on before they are ready as it exposes stability/compatibility concerns for the exported metrics. We were going to have them default-off initially?

Reply:
Yes, no environment variable is fine, but if the plan is to have them off by default initially (enabled via an API) then that SGTM.

Reply:
What about when we decide to stabilize them?  Will they be on by default or off by default at that time?  If they'll be on by default, then there will be a behavior change.  Will there be a gRFC for that change or should this gRFC cover how to de-experimentalize them?

Reply:
I've been thinking about de-experimentalizing metrics as being roughly analogous to how we handle removing env var guards for new features: the gRFC really specifies only how the feature will work when it's enabled by default, even though the expectation is that the feature will start disabled by default.  Just as we don't need a separate gRFC to remove env var guards, I wouldn't think we'd need one to de-experimentalize metrics.

While it is true that gRFCs for features do say something about the env var guards, I think that's necessary because (a) we need to specify the env var name so that we use the same name cross-language and (b) we need to specify exactly which part of the code is guarded by the env var.  Neither of those things are issues with metrics, so I don't think this needs to be called out in the gRFC.  But if either of you would feel more comfortable with some boilerplate that says "all metrics start as experimental and will be de-experimentalized later", I'm fine with that.

The only other difference I see between the two is that the criteria for de-experimentalizing is different: in the case of features, it happens when interop tests pass, whereas for metrics, it will probably happen more when we decide that we want to fully commit to the metric definitions.  But I'm not sure how much that matters.

I'm open to alternative approaches here, but I would prefer to define a process once in A79 and not need a lot of boilerplate in each individual gRFC that defines new metrics.

Reply:
Maybe I'm missing something, though, because I don't see where this gRFC says which of these metrics would be enabled-by-default or disabled-by-default.  This seems like something that should be included.

Or maybe you're saying it doesn't matter?  Like we might take optional metrics and change them to be enabled by default at will, or vice-versa, whenever we want?

IMO for each metric we should call out how we intend to treat it, according to the list here:

https://github.com/grpc/proposal/pull/421/files#diff-58a6eedc2e7a57c026455f8674357915d0943d732e594574cd62bd83dcdd1985R330

Reply:
What I'm saying is that the A79 draft defines what I think is the normal progression for all metrics:

"""
New metrics should start as experimental. Once they are stable, they can optionally be made on-by-default.
"""

I think we should assume that all metrics follow this path without having to repeat what the path is in every individual gRFC that defines new metrics.

I think that we should be able to move a metric from experimental to stable at any time and from off-by-default to on-by-default at any time.  The reverse moves are not necessarily okay, though.

Reply:
So who makes the decision to make a metric on or off by default?  Each language owner?

Reply:
I think we should assume that all metrics are intended to eventually be on by default unless a gRFC explicitly states otherwise.

Reply:
> it will probably happen more when we decide that we want to fully commit to the metric definitions

How would someone implementing the gRFC know if we have committed? As far as I'm concerned, the "Temporary environment variable protection" section should really say when the protection can be removed, which is generally "when interop tests pass." That's even more important here if this is a cross-language decision.

Even if we have a normal process, I do think it is appropriate to mention that things will start experimental and reference that process in the gRFC. Messing this up messes up all implementations, which is to me is why we have the section about the environment variable protection. (I don't care much about the standardized variable name; that is convenient but non-essential. The benefit was "don't pollute new features with broken implementations, which then require design workarounds".)

> I think we should assume that all metrics are intended to eventually be on by default unless a gRFC explicitly states otherwise.

That is definitely not part of A79.

>> Once they are stable, they can optionally be made on-by-default.

That language tells me we should document in the gRFC whether it will be on-by-default.

Reply:
Okay, I've added a section on metric stability.  My expectation is that we will wind up cutting-and-pasting this exact text into all future gRFCs that define new metrics.

Reply:
Okay, given the updated text, for this gRFC we _aren't_ making stable metrics (opposite of what you were saying earlier). That's fine. I'm very glad we added that, because I don't think we'd have done similar things cross-language.